### PR TITLE
postgresql: more cleanup

### PR DESCRIPTION
**Pull request by wolfgangwalther**
## Description of changes

TLDR: This is a continuation of #292993 including those changes which will cause rebuilding the postgresql derivations. Only cleanup, no functional changes.

While working on building PostgreSQL with meson as mentioned in #292993, I had to carefully look at every line in the old builder to see if and how to take this to the meson build. Some of the special cases seem to be not needed anymore, so this PR removes them. Trying to understand the purpose of each of the custom patches, I also cleaned those up along the way with some common formatting and improved naming. Hopefully, this would not only help my own understanding, but other's, too. Finally this PR also contains one more change that is related to the refactor of `this` in the other PR - here, I remove `thisAttr`, too, and make sure that `.tests` will always work correctly, even when overriding the derivation.

~~This merges into the branch of #292993, so please ignore the first 13 or so commits (all starting with "refactor ...") - those are discussed in the other PR.~~ I still split this into a second PR, because the other changes are a pure refactor without causing rebuilds - while those changes there will cause a full rebuild of all postgresql derivations. Thus, I also targeted staging instead of master with this PR here.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Maintainers: @thoughtpolice @danbst @globin @marsam @ivan @Ma27

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Ma27** [comment]
For now, just one request: please don't @ me in commit messages (feel free to reference my GH handle or my name, but without @). The reason is that too many people out there have interesting rebase workflows in their forks that cause notifications in my inbox whenever they push something (had this issue a few times already) and it was always pretty annoying.

**Comment by Ma27** [comment]
Just to make sure it doesn't get lost, from https://github.com/NixOS/nixpkgs/pull/292993#issuecomment-1987201845:

would you mind trying out if adding 2da789a to .git-blame-ignore-revs helps to retain a bit of the history of the previous file?

No idea if that works out, but maybe worth a try?

**Comment by Ma27** [comment]
I think this needs another rebase.

File: nixos/tests/postgresql-wal-receiver.nix [hunk 1]
```diff
@@ -9,111 +10,117 @@ let
...
+      replicationUser = "wal_receiver_user";
+      replicationSlot = "wal_receiver_slot";
+      replicationConn = "postgresql://${replicationUser}@localhost";
+      baseBackupDir = "/tmp/pg_basebackup";
+      walBackupDir = "/tmp/pg_wal";
+      atLeast12 = lib.versionAtLeast pkg.version "12.0";
+
+      recoveryFile = if atLeast12
+          then pkgs.writeTextDir "recovery.signal" ""
+          else pkgs.writeTextDir "recovery.conf" "restore_command = 'cp ${walBackupDir}/%f %p'";
```
Comment (marsam) [inline nixos/tests/postgresql-wal-receiver.nix:31]: nitpick, we no longer have PostgreSQL<12 in nixpkgs, so I think we can remove this branch

**Comment by Ma27** [comment]
> Can I do anything to help the review process?

Not on my end unfortunately, it's mostly an issue of time vs. other stuff I care about. That said, this PR and another larger change I'd really like to see moved forward are pretty high up on my todo list, so I'd like to assure you that I haven't forgotten about this and that I'll do my best to help bringing this forward soonish.

**Review by marsam — APPROVED** [review]
LGTM. Great work, thanks a lot!

**Comment by Ma27** [comment]
Looking good. Would like to run a few tests on my own with it before approving and merging.

Btw cc @yu-re-ka would you mind taking a look at the musl-related patches?

Nice work!

**Comment by Ma27** [comment]
Apologies for the unnecessary ping then!

I thought there was discussion around that, but I didn't find it in this thread, sorry.

**Review by Ma27 — APPROVED** [review]
